crates/core/src/models/block.rs

```diff
@@ -1,15 +1,50 @@
 use async_trait::async_trait;
-use reth_primitives::{Bloom, Bytes, H256, H64, U256};
+use reth_primitives::{BlockId as EthereumBlockId, BlockNumberOrTag, Bloom, Bytes, H256, H64, U256};
 use reth_rpc_types::{Block, BlockTransactions, Header, RichBlock};
-use starknet::core::types::{FieldElement, MaybePendingBlockWithTxHashes, MaybePendingBlockWithTxs, Transaction};
+use starknet::core::types::{
+    BlockId as StarknetBlockId, BlockTag, FieldElement, MaybePendingBlockWithTxHashes, MaybePendingBlockWithTxs,
+    Transaction,
+};
 use starknet::providers::jsonrpc::JsonRpcTransport;
 
 use super::convertible::ConvertibleStarknetBlock;
+use super::felt::Felt252Wrapper;
 use crate::client::api::KakarotEthApi;
 use crate::client::constants::{DIFFICULTY, GAS_LIMIT, GAS_USED, MIX_HASH, NONCE, SIZE, TOTAL_DIFFICULTY};
 use crate::client::errors::EthApiError;
 use crate::client::helpers::starknet_address_to_ethereum_address;
 
+pub struct EthBlockId(EthereumBlockId);
+
+impl EthBlockId {
+    pub fn new(block_id: EthereumBlockId) -> Self {
+        Self(block_id)
+    }
+
+    pub fn to_starknet_block_id<T: JsonRpcTransport>(&self) -> Result<StarknetBlockId, EthApiError<T::Error>> {
+        match self.0 {
+            EthereumBlockId::Hash(hash) => {
+                let hash: Felt252Wrapper = hash.block_hash.try_into()?;
+                Ok(StarknetBlockId::Hash(hash.into()))
+            }
+            EthereumBlockId::Number(block_number_or_tag) => match block_number_or_tag {
+                BlockNumberOrTag::Safe | BlockNumberOrTag::Latest | BlockNumberOrTag::Finalized => {
```

Comment:
here also a opinionated choice -> safe, latest and finalized all link to latest on starknet, might not be v sound!
to be fair -> kakarot is not PoS, so it doesnt even make sense to talk about safe or finalized, we could also fail the conversion from there.